File: registry/scripts/check-r2.pl
#!/usr/bin/env perl
#
# The lexical half of scripts/check-r2.sh (see there, and
# docs/architecture.md, "The cost governor"): every way the Worker can
# obtain an R2 bucket handle (`env.bucket(...)` in any spelling) must
# sit in a function that was reviewed to admit its billable calls
# through the governor first. The allowlist below pins those functions
# with their exact acquisition counts; a new site (or a new acquisition
# inside a pinned function) fails until the reviewer confirms the
# governor admission and re-pins it here. Reads the Rust sources named
# on the command line; prints every violation and exits non-zero if
# there is one.

use strict;
use warnings;
use FindBin;

require "$FindBin::RealBin/lexical.pm";

# (path suffix) => { enclosing fn => sanctioned acquisition count }.
# glue.rs: the four request-path acquisitions are each immediately
# preceded by a governor decide (artifact/read/publish paths), the
# reclaim delete is R2's one free operation, the heal and drain helpers
# admit per call inside, and the cron entry acquires both buckets once
# for the drain. web_glue's source viewer and backup_glue's dump job
# admit before every billable call (docs/architecture.md).
my %allow = (
    'src/glue.rs' => {
        artifact_response           => 1,
        charged_blob_read           => 1,
        persist_new_version         => 1,
        replace_rejected_version    => 1,
        delete_blob_if_unreferenced => 1,
        heal_blobs_on_retry         => 1,
        drain_backup_queue          => 2,
    },
    'src/web_glue.rs'    => { package_source   => 1 },
    'src/backup_glue.rs' => { run_nightly_dump => 1 },
);

my $fail = 0;
for my $file (@ARGV) {
    open my $handle, '<', $file or die "$file: $!\n";
    my $source = blank_comments_and_strings(do { local $/; <$handle> });
    my ($suffix) = grep { $file =~ m{(?:^|/)\Q$_\E$} } keys %allow;
    my $sanctioned = $suffix ? $allow{$suffix} : {};

    # Every `fn name` position, for attributing a call site to its
    # nearest preceding function. A closure inside a function
    # attributes to that function; good enough for a pin whose job is
    # to force review, not to prove admission.
    my @fns;
    while ($source =~ m{\bfn\s+(?:r\#)?([A-Za-z0-9_]+)}g) {
        push @fns, [ $-[0], $1 ];
    }

    my %seen;
    # Any spelling that can reach Env::bucket - `.bucket(`,
    # `::bucket(`, `r#bucket`, split across lines or comments. The
    # word boundary keeps `bucket_from_columns` out; the lookahead for
    # `(` keeps field access (`auth.bucket`) out.
    while ($source =~ m{ [.:] \s* (?: r\# )? bucket \b (?= \s* \( ) }gsx) {
        my $offset = $-[0];
        my ($enclosing) = grep { $_->[0] <= $offset } reverse @fns;
        my $fn = $enclosing ? $enclosing->[1] : '(no enclosing fn)';
        $seen{$fn}++;
        next if exists $sanctioned->{$fn} && $seen{$fn} <= $sanctioned->{$fn};
        my $line = 1 + (substr($source, 0, $offset) =~ tr/\n//);
        print "$file:$line: unsanctioned R2 bucket acquisition in $fn - "
            . "prove the governor admission and pin it in scripts/check-r2.pl\n";
        $fail = 1;
    }

    # A path-form method item (`Env::bucket` with no call parens) is an
    # alias that would launder every later acquisition past the scan
    # above, so creating one is itself a violation. The dotted form
    # needs no twin check: `env.bucket` without parens is not valid
    # Rust for a method value, and plain field access is another name.
    while ($source =~ m{ :: \s* (?: r\# )? bucket \b (?! \s* \( ) }gsx) {
        my $line = 1 + (substr($source, 0, $-[0]) =~ tr/\n//);
        print "$file:$line: R2 bucket method alias (path form without a call); "
            . "call it directly in a pinned function instead\n";
        $fail = 1;
    }

    # worker's generic `env.get_binding::<T>(name)` can also yield a
    # Bucket without the `bucket` token appearing at all, and an
    # unchecked JS cast can conjure one from any JsValue. Nothing in
    # this Worker needs either, so both are banned outright. (Reflect
    # over the raw env object could still do it - that is deliberate
    # evasion, which is code review's job, not a tripwire's.)
    while ($source =~ m{ (?: [.:] \s* )? (?: r\# )? (get_binding|unchecked_into) \b }gsx) {
        my $line = 1 + (substr($source, 0, $-[0]) =~ tr/\n//);
        print "$file:$line: $1 sidesteps the typed accessors "
            . "(an R2 handle without the bucket token); use the typed "
            . "Env method in a pinned function\n";
        $fail = 1;
    }

    # A pinned function that still exists but no longer acquires its
    # bucket means the seam moved; the pin must move with it or it
    # stops guarding. A pin whose function is gone entirely is left to
    # the review that removed the function.
    my %defined = map { $_->[1] => 1 } @fns;
    for my $fn (sort keys %{$sanctioned}) {
        next unless $defined{$fn};
        next if ($seen{$fn} // 0) == $sanctioned->{$fn};
        next if ($seen{$fn} // 0) > $sanctioned->{$fn};    # already reported
        print "$file: $fn is pinned for $sanctioned->{$fn} acquisition(s) "
            . "but has " . ($seen{$fn} // 0) . "; update scripts/check-r2.pl\n";
        $fail = 1;
    }
}
exit($fail ? 1 : 0);

File: crates/xtask-registry-guard/tests/check_r2.rs
//! Regression cases for the R2 acquisition guard (see
//! `registry/docs/architecture.md`, "The cost governor"): the guard runs
//! against a scratch tree whose `src/` holds synthetic call sites, so
//! every way a bucket handle could be acquired outside the pinned
//! governor-admitting functions - a new function, a second acquisition
//! inside a pinned one, the UFCS and raw-identifier spellings, the name
//! split from its receiver by a comment - stays caught, and the shapes
//! that are not acquisitions at all stay accepted. An untested guard is
//! the one that rots.

use assert_cmd::Command;
use predicates::str::contains;
use std::fs;

use xtask_registry_guard::{r2, registry_dir};

/// A scratch registry tree whose `src/<file>` holds `call_site`.
fn scratch(file: &str, call_site: &str) -> assert_fs::TempDir {
    let dir = assert_fs::TempDir::new().expect("scratch tree");
    let path = dir.path().join("src").join(file);
    fs::create_dir_all(path.parent().expect("src parent")).expect("create scratch src/");
    fs::write(path, call_site).expect("write the call site");
    dir
}

/// Runs the guard over a scratch tree containing `call_site` at
/// `src/<file>`; `true` means the guard accepted it.
fn guard_accepts_in(file: &str, call_site: &str) -> bool {
    let dir = scratch(file, call_site);
    r2::check(dir.path()).expect("run the guard").is_empty()
}

/// A pinned function holding exactly its sanctioned acquisitions - and
/// the neighboring shapes that are not acquisitions at all - must pass,
/// or the guard would block ordinary work.
#[test]
fn the_canonical_call_sites_pass() {
    let accepted = guard_accepts_in(
        "glue/read.rs",
        concat!(
            "async fn artifact_response(env: &Env) -> worker::Result<Response> {\n",
            "    let Some(object) = env.bucket(\"BLOBS\")?.get(&key).execute().await? else {\n",
            "        return not_found();\n",
            "    };\n",
            "}\n",
            // Field access is not a call, a lookalike name is not the
            // method, and a comment describing one is not code.
            "fn bucket_from_columns(auth: &AuthContext) -> Option<quota::Bucket> {\n",
            "    if auth.bucket.is_some() { return None; }\n",
            "    // The call sites go through env.bucket(\"BLOBS\") after a decide.\n",
            "    let doc = r#\"{\"call\":\"env.bucket(x)\"}\"#;\n",
            "    read_bucket(db, &auth.token_id)\n",
            "}\n",
        ),
    );
    assert!(accepted, "the guard rejected the canonical call sites");
    // The queue drain's double acquisition is pinned under
    // backup_glue.rs, where the drain lives.
    let accepted = guard_accepts_in(
        "backup_glue.rs",
        concat!(
            "async fn drain_backup_queue(env: &Env) {\n",
            "    let (Ok(db), Ok(blobs), Ok(backup)) =\n",
            "        (env.d1(\"DB\"), env.bucket(\"BLOBS\"), env.bucket(\"BACKUP\"));\n",
            "}\n",
        ),
    );
    assert!(accepted, "the guard rejected the backup_glue drain");
}

#[test]
fn unsanctioned_acquisitions_are_caught() {
    // Each is a distinct way a bucket handle could be acquired outside
    // the pinned seam.
    let cases: &[(&str, &str)] = &[
        (
            "new_function",
            "async fn sneaky_reader(env: &Env) { let b = env.bucket(\"BLOBS\")?; }",
        ),
        (
            // The pin is a count: a second acquisition inside a
            // sanctioned function is a new seam to review.
            "second_acquisition_in_a_pinned_fn",
            concat!(
                "async fn artifact_response(env: &Env) {\n",
                "    let a = env.bucket(\"BLOBS\")?;\n",
                "    let b = env.bucket(\"BACKUP\")?;\n",
                "}\n",
            ),
        ),
        (
            "ufcs",
            "fn f(env: &Env) { let b = worker::Env::bucket(env, \"BLOBS\"); }",
        ),
        (
            "raw_identifier",
            "fn f(env: &Env) { let b = env.r#bucket(\"BLOBS\"); }",
        ),
        (
            "comment_between_receiver_and_name",
            "fn f(env: &Env) { let b = env./* sneaky */bucket(\"BLOBS\"); }",
        ),
        (
            // a line-oriented match would miss this; the scan must not be.
            "comment_between_name_and_paren_across_lines",
            "fn f(env: &Env) {\n    let b = env.\n/* explanation */\nbucket\n(\"BLOBS\");\n}",
        ),
        (
            // A `//` inside a string starts no comment: the call after
            // it on the same line must still be seen.
            "after_a_url_string",
            "fn f(env: &Env) { let u = \"https://api.cloudflare.com\"; let b = env.bucket(\"BLOBS\"); }",
        ),
        (
            "outside_any_fn",
            "static B: () = { env.bucket(\"BLOBS\") };",
        ),
        (
            // A path-form method item aliases the method; every later
            // call through the alias would evade the call scan.
            "method_item_alias",
            "fn f(env: &Env) { let acquire = worker::Env::bucket; acquire(env, \"BLOBS\"); }",
        ),
        (
            // The generic binding accessor yields a Bucket without the
            // `bucket` token ever appearing.
            "generic_get_binding",
            "fn f(env: &Env) { let b: worker::Bucket = env.get_binding(\"BLOBS\").unwrap(); }",
        ),
        (
            // So does an unchecked JS cast over the raw env object.
            "unchecked_cast",
            "fn f(v: JsValue) { let b = v.unchecked_into::<worker::Bucket>(); }",
        ),
    ];
    let escaped: Vec<&str> = cases
        .iter()
        .filter(|(_, call_site)| guard_accepts_in("glue/read.rs", call_site))
        .map(|(name, _)| *name)
        .collect();
    assert!(
        escaped.is_empty(),
        "the guard accepted an unsanctioned R2 acquisition: {escaped:?}"
    );
}

/// The pins are file-scoped: a sanctioned glue/read.rs function does not
/// sanction the same name elsewhere, and a pinned function that no
/// longer acquires its bucket is drift the pin must follow.
#[test]
fn the_pins_are_file_scoped_and_track_drift() {
    assert!(!guard_accepts_in(
        "verify.rs",
        "async fn artifact_response(env: &Env) { let b = env.bucket(\"BLOBS\")?; }",
    ));
    assert!(!guard_accepts_in(
        "glue/read.rs",
        "async fn artifact_response(env: &Env) { serve_from_cache(env).await }",
    ));
}

/// The drift diagnostic names the function, both counts, and where the
/// pin lives - a reviewer has to be able to act on it.
#[test]
fn a_drifted_pin_names_its_counts() {
    let dir = scratch(
        "backup_glue.rs",
        "async fn drain_backup_queue(env: &Env) { let b = env.bucket(\"BLOBS\"); }",
    );
    let violations = r2::check(dir.path()).expect("run the guard");
    assert_eq!(
        violations,
        vec![
            "src/backup_glue.rs: drain_backup_queue is pinned for 2 acquisition(s) \
             but has 1; update crates/xtask-registry-guard/src/r2.rs"
        ]
    );
}

/// Several drifted pins in one file come out sorted by function name.
#[test]
fn drifted_pins_are_reported_in_name_order() {
    let dir = scratch(
        "glue/bearer/package.rs",
        // Chosen so the allowlist order and the alphabetical order
        // disagree: `persist_new_revision` is pinned before
        // `heal_blobs_on_retry` but sorts after it.
        "fn persist_new_revision() {}\nfn heal_blobs_on_retry() {}\n",
    );
    let violations = r2::check(dir.path()).expect("run the guard");
    let names: Vec<&str> = violations
        .iter()
        .map(|line| line.split_whitespace().nth(1).expect("the function name"))
        .collect();
    assert_eq!(names, vec!["heal_blobs_on_retry", "persist_new_revision"]);
}

/// An acquisition outside every function is attributed to no function
/// rather than silently landing in the previous one.
#[test]
fn an_acquisition_outside_any_function_says_so() {
    let dir = scratch("verify.rs", "static B: () = { env.bucket(\"BLOBS\") };");
    let violations = r2::check(dir.path()).expect("run the guard");
    assert_eq!(
        violations,
        vec![
            "src/verify.rs:1: unsanctioned R2 bucket acquisition in (no enclosing fn) - \
             prove the governor admission and pin it in crates/xtask-registry-guard/src/r2.rs"
        ]
    );
}

/// The committed Worker sources pass.
#[test]
fn the_committed_worker_sources_pass() {
    let violations = r2::check(&registry_dir()).expect("run the guard");
    assert!(
        violations.is_empty(),
        "the committed Worker sources: {violations:?}"
    );
}

/// The binary reports violations on stdout, names the remedy on stderr,
/// and exits non-zero - the contract CI depends on.
#[test]
fn the_binary_reports_and_exits_non_zero() {
    let dir = scratch("verify.rs", "fn f(env: &Env) { env.bucket(\"BLOBS\"); }");
    Command::new(env!("CARGO_BIN_EXE_xtask-registry-guard"))
        .args(["check-r2", "--registry-dir"])
        .arg(dir.path())
        .assert()
        .failure()
        .stdout(contains("unsanctioned R2 bucket acquisition in f"))
        .stderr(contains("outside the pinned"));
}
